replication/AmbryReplication.tla

```diff
@@ -0,0 +1,328 @@
+        /\ blobId < blobIdCtr
+        /\ replicaStates[r1] = CanTakeWrite
+        /\ replicaStates[r2] = CanTakeWrite
+        /\ [ blobid |-> blobId, blobmessage |-> PUT ] \in storedMessagesPerReplica[r1]
+        /\ [ blobid |-> blobId, blobmessage |-> DELETE ] \notin storedMessagesPerReplica[r1]
+        /\ [ blobid |-> blobId, blobmessage |-> PUT ] \in storedMessagesPerReplica[r2]
+        /\ [ blobid |-> blobId, blobmessage |-> DELETE ] \notin storedMessagesPerReplica[r2]
+        /\ [ blobid |-> blobId, blobmessage |-> UPDATE_TTL ] \notin storedMessagesPerReplica[r2]
+        /\ [ blobid |-> blobId, blobmessage |-> UPDATE_TTL ] \notin storedMessagesPerReplica[r1]
+        /\ storedMessagesPerReplica' = [rr \in DOMAIN storedMessagesPerReplica |-> 
+            IF ((rr = r1) \/ (rr = r2)) THEN
+                storedMessagesPerReplica[rr] \union { [ blobid |-> blobId, blobmessage |-> UPDATE_TTL ] }
+            ELSE
+                storedMessagesPerReplica[rr]
+            ]
+        /\ updateTtlCtr' = updateTtlCtr + 1
+    /\ UNCHANGED <<replicaStates, blobIdCtr, inflightReplicationMessages, deleteCtr, inflightReplicationRequest, inflightReplicationResponse, recvReplicationReqSet, sendRecplicationSet, recvReplicationRespSet>>
+
+(* Bring one replica up and one replica down *)
+ReplicaStateChange ==
+    \E m, n \in Replicas:
+        \/ ReplicaDown(n)
+        \/ ReplicaUp(m)
+
+(* Ambry uses a pull model for replication.
+ * In SendReplicationRequest action, a replica asks one of its peers to give the set of data the peer has.
+ * This action just enqueues the request to peer in "inflightReplicationRequest".
+ *)
+SendReplicationRequest(src) ==
+    /\ replicaStates[src] = CanTakeWrite \* We want replica asking for replication pull to be writable.
+    /\ \E tgt \in Replicas:
+        /\ src # tgt
+        /\ inflightReplicationRequest' = [inflightReplicationRequest EXCEPT ![tgt] = inflightReplicationRequest[tgt] \union {src} ]
+    /\ IF AllBlobOpsDone THEN 
+           sendRecplicationSet' = sendRecplicationSet \union {src}
+       ELSE
+           sendRecplicationSet' = sendRecplicationSet
+    /\ UNCHANGED <<replicaStates, blobIdCtr, storedMessagesPerReplica, deleteCtr, updateTtlCtr, inflightReplicationMessages, inflightReplicationResponse, recvReplicationReqSet, recvReplicationRespSet>>
+
+(* Ambry uses a pull model for replication.
+ * In RecvReplicationRequest action, a replica recieves replication request from a peer and sends all of its storage log.
+ * This action enqueues the entire storage log "inflightReplicationResponse".
+ *)
+RecvReplicationRequest(r) ==
+    /\ inflightReplicationRequest[r] # {}
+    /\ storedMessagesPerReplica[r] # {} \* there should be something to replicate.
+    /\ \E src \in inflightReplicationRequest[r]:
+        /\ inflightReplicationResponse' = [inflightReplicationResponse EXCEPT ![src] = inflightReplicationResponse[src] \union {[r |-> r, messages |-> storedMessagesPerReplica[r]]} ]
```

Comment:
nit: it looks like inflightReplicationResponse is a map where the index of the map corresponds to a replicaId.  In which case, do we need r |-> r?  We'd get that from the index.  If it's for clarity sake, then I'd recommending changing the field name from r to replicaId

Reply:
Actually nm, I see that actually this is keyed on source replica and the message contains the id of the destination (where src is the sender of the response and r will retrieve it).  So in that case, keep r, but maybe tweak the member name to make it more clear.